src/action.c

```diff
@@ -1301,4 +1301,18 @@ actions_run(struct view *activator, struct server *server,
 				" This is a BUG. Please report.", action->type);
 		}
 	}
+
+	/*
+	 * We send wl_keyboard.{leave,enter} to the clint when we have sent
+	 * key-press events to it since the last wl_keyboard.leave event.
+	 * This prevents firefox to show its menu bar when a keybind with
+	 * Alt key is executed.
+	 */
+	struct seat *seat = &server->seat;
+	if (seat->modifier_press_sent) {
```

Comment:
This is only necessary for keyboard driven actions, right? I think actions_run() is also called on every mouse click in the default config.

Reply:
Well, sending `wl_keyboard.{leave,enter}` also prevents firefox to show menu bar with the default Alt-Drag mousebind. You can see the current problem by running labwc with the default config and dragging firefox window with Alt key pressed.

But I agree that sending those events on every mouse click (by `Focus` & `Raise` mousebinds) doesn't feel right. Perhaps we should send those events only with keybinds and we can clear keyboard focus separately in `interactive_begin()`.

Reply:
That sounds like a good approach 👍

Reply:
Hmm... sorry, I was totally wrong in this part:

> But I agree that sending those events on every mouse click (by Focus & Raise mousebinds) doesn't feel right. 

In this PR, leave & enter events are only sent when any modifier is pressed, so the default Focus & Raise mousebind doesn't trigger leave & enter events.

Furthermore, I came up with another case where sending leave & enter events also after mousebinds is beneficial; consider following mousebinds:

```xml
    <context name="All">
      <mousebind action="Scroll" direction="A-Up">
        <action name="ZoomIn" />
      </mousebind>
      <mousebind action="Scroll" direction="A-Down">
        <action name="ZoomOut" />
      </mousebind>
    </context>
```

In this case, I guess the user doesn't want firefox to show menu bar after updating magnification rate.

Therefore, now I think putting the logic for sending leave & enter events in `actions_run()` is a better idea.

Reply:
That's a good thought. It still seems like we should send the enter/leave when the modifier is part of a compositor action (like the magnifier thing) but not when it's a client action (like Ctrl+Scroll to zoom). I'm not sure of the best mechanism to do that.

Reply:
Sorry, I don't get what you mean. In this PR, leave & enter events are sent when a keybind or mousebind is executed while a modifier has been pressed. So by default, we never send leave & enter when user presses Ctrl+Scroll. Of course, when the user sets mousebinds like `<mousebind action="Scroll" direction="C-Up">` it may interfere with the client-side action, but I think that can be helped and the user should not set such a mousebind anyway.

My current concerns with this PR are:
- It deviates from what other compositors like KWin do. So _perhaps_ it causes issues with applications that don't expect the behavior in this PR.
- It sometimes sends unnecessary leave & enter events. For example, we practically don't need to send those events when launching a new terminal window because a leave event is sent to the original window anyway when moving the keyboard focus to the new window. But in this PR, we send leave -> enter -> leave events to the original window, which feels a bit strange.

But overall, I think this PR does its job well.

Reply:
Ctrl+Scroll may have been a bad example. In my setup I like to have Ctrl+Click of any mouse button bound to "Focus" and "Raise". But I don't want every Ctrl+Click to send keyboard enter & leave events.

Reply:
OK. I see sending leave & enter events only for keybinds is preferable on your setup.

However, now I wonder if we have a good reason to differentiate keybinds vs mousebinds in regard to whether to send the leave & enter events, as both keybinds and mousebinds can unexpectedly open firefox's menu bar.

The more I consider this problem, the more I think we should take the option (2) in #2448: just send a leave event on window-switcher activation. This only works for the default Alt-Tab and will reintroduce the firefox's menu bar issue with some existing keybinds like:

```xml
    <keybind key="A-Left">
      <action name="MoveToEdge" direction="left" />
    </keybind>
```

But Kwin also has this problem and I think it's a much more minor issue than that with Alt-Tab keybind. As we advertise labwc as a lightweight and simple compositor, I prefer this way.

I'll open another PR with this approach.

Reply:
I'm fine with that. Thanks for all the work on this - it's tricky to get right.

I think the leave/enter approach could also still work, if we sorted out what actions it should or shouldn't trigger for. But maybe it needs more thought.